File: leetcode-cn/0229.0_Majority_Element_II.py
'''
摩尔投票法
T: O(2N)
S: O(1)

执行用时：48 ms, 在所有 Python3 提交中击败了26.45% 的用户
内存消耗：16 MB, 在所有 Python3 提交中击败了5.13% 的用户
通过测试用例：82 / 82
'''
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        element1 = element2 = 0
        vote1 = vote2 = 0
        for x in nums:
            if vote1 > 0 and element1 == x:
                vote1 += 1
            elif vote2 > 0 and element2 == x:
                vote2 += 1
            elif vote1 == 0:
                element1 = x
                vote1 = 1
            elif vote2 == 0:
                element2 = x
                vote2 = 1
            else:
                vote1 -= 1
                vote2 -= 1
        
        cnt1 = cnt2 = 0
        for x in nums:
            if x == element1:
                cnt1 += 1
            elif x == element2:
                cnt2 += 1
        ans = []
        if cnt1 > len(nums) // 3:
            ans.append(element1)
        if cnt2 > len(nums) // 3:
            ans.append(element2)
        return ans
        

File: cpp/0229.0_Majority_Element_II.cpp
/*
执行用时：12 ms, 在所有 C++ 提交中击败了68.66% 的用户
内存消耗：15.3 MB, 在所有 C++ 提交中击败了82.08% 的用户
通过测试用例：82 / 82
*/

class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        vector<int> ans;
        int element1 = 0, element2 = 0;
        int vote1 = 0, vote2 = 0;
        for (int i = 0; i < nums.size(); ++i) {
            if (vote1 > 0 && nums[i] == element1) {
                vote1++;
            } else if (vote2 > 0 and nums[i] == element2) {
                vote2++;
            } else if (vote1 == 0) {
                element1 = nums[i];
                vote1 = 1;
            } else if (vote2 == 0) {
                element2 = nums[i];
                vote2 = 1;
            } else {
                vote1--;
                vote2--;
            }
        }

        int cnt1 = 0, cnt2 = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] == element1) {
                cnt1++;
            } else if (nums[i] == element2) {
                cnt2++;
            }
        }
        if (cnt1 > nums.size() / 3) {
            ans.push_back(element1);
        }
        if (cnt2 > nums.size() / 3) {
            ans.push_back(element2);
        }
        return ans;
    }
};
